### PR TITLE
ENH: DDL support for Spark backend

### PR DESCRIPTION
**Pull request by amroid**

### CODE REVIEW

File: ibis/spark/ddl.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
+import re
```
Comment (cpcloud) [inline ibis/spark/ddl.py]: Why not reuse the code instead of copying it?

File: ibis/spark/ddl.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+
+def _format_properties(props):
+    tokens = []
+    for k, v in sorted(props.items()):
+        tokens.append("  '{}'='{}'".format(k, v))
+
+    return '(\n{}\n)'.format(',\n'.join(tokens))
+
+
+class CreateTable(CreateDDL):
```
Comment (cpcloud) [inline ibis/spark/ddl.py]: For example, why not inherit from the impala DDL's version of this class?

File: ibis/spark/ddl.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+            table_name,
+            database=database,
+            format=format,
+            can_exist=can_exist,
+            path=path,
+        )
+        self.select = select
+
+    @property
+    def _pieces(self):
```
Comment (cpcloud) [inline ibis/spark/ddl.py]: You don't need to duplicate this method because `self.partition` is `None` by default.

File: ibis/spark/ddl.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        super().__init__(
+            table_name,
+            select,
+            database=database,
+            can_exist=or_replace,
+        )
+        self.temporary = temporary
+
+    @property
+    def _pieces(self):
```
Comment (cpcloud) [inline ibis/spark/ddl.py]: This doesn't appear to be necessary.

File: ibis/spark/ddl.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    return '{} {}'.format(
+        quote_identifier(name, force=True), _type_to_sql_string(t)
+    )
+
+
+class CreateDatabase(impala_ddl.CreateDatabase):
+    pass
+
+
+class DropObject(impala_ddl.DropObject):
```
Comment (cpcloud) [inline ibis/spark/ddl.py]: Is there a reason to define this? It looks like it's only used in `DropDatabase`. Why not just directly subclass `impala_ddl.DropObject` in the definition of `DropDatabase`?

File: ibis/spark/ddl.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        self.cascade = cascade
+
+    def _object_name(self):
+        return self.name
+
+    def compile(self):
+        if self.cascade:
+            return '{} CASCADE'.format(super().compile())
+        else:
+            return super().compile()
```
Comment (cpcloud) [inline ibis/spark/ddl.py]: Since you're calling `super().compile()` regardless of the branch taken let's put the result of calling that method into a variable.

File: ibis/spark/ddl.py (hunk 5 above)
Comment (cpcloud) [reply ibis/spark/ddl.py]: Also, it looks like `DropObject` is a parent of `DropTable` in `impala_ddl`. Did you mean to do that in this file as well?

File: ibis/spark/ddl.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+class CreateView(CTAS):
+
+    """Create a view"""
+
+    def __init__(
+        self,
+        table_name,
+        select,
+        database=None,
+        or_replace=True,
```
Comment (cpcloud) [inline ibis/spark/ddl.py]: Why did you change the API here? Shouldn't this be `can_exist`?

File: ibis/spark/tests/test_ddl_compilation.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+)"""
+    assert result == expected
+
+
+@pytest.fixture
+def expr(t):
+    return t[t.bigint_col > 0]
+
+
+def test_create_external_table_as(mockcon):
```
Comment (cpcloud) [inline ibis/spark/tests/test_ddl_compilation.py]: I thought we weren't implementing external tables.

File: ibis/spark/tests/test_ddl_compilation.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+        table_name,
+        select,
+        database=database,
+        format=format,
+        can_exist=can_exist,
+    )
+    return statement
+
+
+def _get_select(expr, context):
```
Comment (cpcloud) [inline ibis/spark/tests/test_ddl_compilation.py]: The context is never used where this function is called, so do we really need this function? It's not clear to me what's being saved here.

File: ibis/spark/tests/test_ddl_compilation.py (hunk 8 above)
Comment (cpcloud) [reply ibis/spark/tests/test_ddl_compilation.py]: Can you remove these tests for now?

**Comment by cpcloud** [comment]
@amroid Looks like the CI is failing

File: ibis/spark/ddl.py (hunk 4 above)
Comment (amroid) [reply ibis/spark/ddl.py]: I think it's necessary to prevent `self._storage()` and `self._location()` from getting called by `CTAS._pieces`? Since views don't have any storage format or external location.
